### DISCUSSION TITLE
Need help building on windows.

### DISCUSSION THREAD

**Discussion by WuChanGG** [zmwangx/rust-ffmpeg#43] [Q&A]
I read the previous issues regarding building in windows, and read the wiki too, but I still got the `  = note: LINK : fatal error LNK1181: cannot open input file 'avutil.lib'` error despite having the correct FFMPEG_DIR value. I was wondering if there was more input about this. Alternatively I could try to develop with WSL since I'm having a lot of problems finding a video decoder for windows.

**Comment by zmwangx** [comment]
Need a lot more detailed info to help you. OS version, how FFmpeg is installed, exact series of commands to set up your build directory and lead to the error, etc.

**Comment by WuChanGG** [comment]
Windows x64, version 2004, build 19041.685.
FFMPEG is installed by the instructions of the wiki for MSVC toolchain
```
Install FFmpeg (complete with headers) through any means, e.g. downloading a pre-built "full_build-shared" version from https://ffmpeg.org/download.html. See FFMPEG_DIR to the directory containing include and lib.
```
I'm cloning the repository, and running `cargo build` on the root directory seems to work, but trying to run any example e.g. `cargo run --example dump-frames C:\Videos\1280.mp4` gives the following logs 
[example_output.txt](https://github.com/zmwangx/rust-ffmpeg/files/5732550/example_output.txt)
Please let me know if you need more info

**Reply by zmwangx** [reply]
You sure `avutil.lib` and other libs are present under `$FFMPEG_DIR\lib`?

Also, give me the output of

    cargo build -vv --example dump-frames

It should show the full rustc invocation, and at the end or close to the end you should find a linker flag specifying the library lookup path, like

    -L native=C:\Users\zmwang\Projects\rust-ffmpeg\ffmpeg\lib

where all the libs should be under that directory. If there's no such flag then some configuration is amiss. Which is why I asked for the exact series of commands, not natural language descriptions of them.

**Comment by WuChanGG** [comment]
This is the output of the command `cargo build -vv --example dump-frames`, thanks for your time, and sorry for posting not enough info, my bad. and yes the libraries are present in `FFMPEG_DIR/lib`

```
PS H:\RustProjects\ff\rust-ffmpeg> cargo build -vv --example dump-frames
       Fresh version_check v0.9.2
       Fresh glob v0.3.0
       Fresh cc v1.0.66
       Fresh unicode-xid v0.2.1  
       Fresh regex-syntax v0.6.21     
       Fresh shlex v0.1.1
       Fresh lazycell v1.3.0
       Fresh peeking_take_while v0.1.2
       Fresh rustc-hash v1.1.0        
       Fresh cfg-if v0.1.10
       Fresh lazy_static v1.4.0       
       Fresh pkg-config v0.3.19       
       Fresh vcpkg v0.2.11
       Fresh regex v1.4.2
       Fresh winapi v0.3.9
       Fresh memchr v2.3.4
       Fresh libc v0.2.81
       Fresh proc-macro2 v1.0.24
       Fresh bitflags v1.2.1  
       Fresh nom v5.1.2       
       Fresh libloading v0.5.2
       Fresh quote v1.0.8
       Fresh num_cpus v1.13.0
       Fresh cexpr v0.4.0
       Fresh clang-sys v0.29.3
       Fresh bindgen v0.54.0
   Compiling ffmpeg-sys-next v4.3.5
     Running `H:\RustProjects\ff\rust-ffmpeg\target\debug\build\ffmpeg-sys-next-aec6b64fe5535b58\build-script-build`
[ffmpeg-sys-next 4.3.5] cargo:rustc-link-search=native=H:\ffmpeg-n4.3.1-26-gca55240b8c-win64-gpl-shared-4.3;\lib
[ffmpeg-sys-next 4.3.5] cargo:rustc-link-lib=dylib=avcodec
[ffmpeg-sys-next 4.3.5] cargo:rustc-link-lib=dylib=avdevice
[ffmpeg-sys-next 4.3.5] cargo:rustc-link-lib=dylib=avfilter
[ffmpeg-sys-next 4.3.5] cargo:rustc-link-lib=dylib=avformat
[ffmpeg-sys-next 4.3.5] cargo:rustc-link-lib=dylib=avutil
[ffmpeg-sys-next 4.3.5] cargo:rustc-link-lib=dylib=swresample
[ffmpeg-sys-next 4.3.5] cargo:rustc-link-lib=dylib=swscale
[ffmpeg-sys-next 4.3.5] OPT_LEVEL = Some("0")
[ffmpeg-sys-next 4.3.5] HOST = Some("x86_64-pc-windows-msvc")
[ffmpeg-sys-next 4.3.5] CC_x86_64-pc-windows-msvc = None
[ffmpeg-sys-next 4.3.5] CC_x86_64_pc_windows_msvc = None
[ffmpeg-sys-next 4.3.5] HOST_CC = None
[ffmpeg-sys-next 4.3.5] CC = None
[ffmpeg-sys-next 4.3.5] CFLAGS_x86_64-pc-windows-msvc = None
[ffmpeg-sys-next 4.3.5] CFLAGS_x86_64_pc_windows_msvc = None
[ffmpeg-sys-next 4.3.5] HOST_CFLAGS = None
[ffmpeg-sys-next 4.3.5] CFLAGS = None
[ffmpeg-sys-next 4.3.5] CRATE_CC_NO_DEFAULTS = None
[ffmpeg-sys-next 4.3.5] CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
[ffmpeg-sys-next 4.3.5] DEBUG = Some("true")
[ffmpeg-sys-next 4.3.5] cl : Command line warning D9035 : option 'o' has been deprecated and will be removed in a future release
[ffmpeg-sys-next 4.3.5] check.c
[ffmpeg-sys-next 4.3.5] check.c(3): fatal error C1083: Cannot open include file: 'libavutil/avutil.h': No such file or directory
[ffmpeg-sys-next 4.3.5] thread 'main' panicked at 'Compile failed', C:\Users\Wu\.cargo\registry\src\github.com-1ecc6299db9ec823\ffmpeg-sys-next-4.3.5\build.rs:492:9
[ffmpeg-sys-next 4.3.5] stack backtrace:
[ffmpeg-sys-next 4.3.5]    0: std::panicking::begin_panic<str*>
[ffmpeg-sys-next 4.3.5]              at C:\Users\Wu\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\panicking.rs:497
[ffmpeg-sys-next 4.3.5]    1: build_script_build::check_features
[ffmpeg-sys-next 4.3.5]              at .\build.rs:492
[ffmpeg-sys-next 4.3.5]    2: build_script_build::main
[ffmpeg-sys-next 4.3.5]              at .\build.rs:751
[ffmpeg-sys-next 4.3.5]    3: core::ops::function::FnOnce::call_once<fn(),tuple<>>
[ffmpeg-sys-next 4.3.5]              at C:\Users\Wu\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\ops\function.rs:227
[ffmpeg-sys-next 4.3.5] note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
error: failed to run custom build command for `ffmpeg-sys-next v4.3.5`

Caused by:
  process didn't exit successfully: `H:\RustProjects\ff\rust-ffmpeg\target\debug\build\ffmpeg-sys-next-aec6b64fe5535b58\build-script-build` (exit code: 101)
  --- stdout
  cargo:rustc-link-search=native=H:\ffmpeg-n4.3.1-26-gca55240b8c-win64-gpl-shared-4.3;\lib
  cargo:rustc-link-lib=dylib=avcodec
  cargo:rustc-link-lib=dylib=avdevice
  cargo:rustc-link-lib=dylib=avfilter
  cargo:rustc-link-lib=dylib=avformat
  cargo:rustc-link-lib=dylib=avutil
  cargo:rustc-link-lib=dylib=swresample
  cargo:rustc-link-lib=dylib=swscale
  OPT_LEVEL = Some("0")
  HOST = Some("x86_64-pc-windows-msvc")
  CC_x86_64-pc-windows-msvc = None
  CC_x86_64_pc_windows_msvc = None
  HOST_CC = None
  CC = None
  CFLAGS_x86_64-pc-windows-msvc = None
  CFLAGS_x86_64_pc_windows_msvc = None
  HOST_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  DEBUG = Some("true")
  check.c
  check.c(3): fatal error C1083: Cannot open include file: 'libavutil/avutil.h': No such file or directory

  --- stderr
  cl : Command line warning D9035 : option 'o' has been deprecated and will be removed in a future release
  thread 'main' panicked at 'Compile failed', C:\Users\Wu\.cargo\registry\src\github.com-1ecc6299db9ec823\ffmpeg-sys-next-4.3.5\build.rs:492:9
  stack backtrace:
     0: std::panicking::begin_panic<str*>
               at C:\Users\Wu\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\panicking.rs:497
     1: build_script_build::check_features
               at .\build.rs:492
     2: build_script_build::main
               at .\build.rs:751
     3: core::ops::function::FnOnce::call_once<fn(),tuple<>>
               at C:\Users\Wu\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\core\src\ops\function.rs:227
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
PS H:\RustProjects\ff\rust-ffmpeg>
```

**Comment by zmwangx** [comment]
So I'm confused, you said `cargo build` would succeed and `cargo run` would fail at linker stage, but this time it doesn't even get past compile stage, not being able to find the header. I suspect `FFMPEG_DIR` isn't set correctly. What shell is this (cmd? PowerShell?) and how are you setting `FFMPEG_DIR`?  Still I don't see your exact series of commands.

Let me be clear why I ask for the *exact series of commands*, not natural language descriptions of them: often times when someone say "I did X, then Y, and Y failed, here's the output of Y", implying fault somewhere in Y, what they actually mean is "I think I did X, then Y, ..." , and in fact they failed to do X properly, so the failure was in X which is apparent once the exact steps are provided.

Basically, any step could have gone wrong, so you need to tell me how you did each step *exactly*, not just the final step.

**Comment by WuChanGG** [comment]
OKAY thanks a lot for your time, I was setting the FFMPEG_DIR and the LIBCLANG_PATH environment variables by using the windows GUI for environment variables, it never occurred to me that they weren't working (I don't know why, I restart my pc every time I set new environment variables). So I opened the project in VSCODE and set the environment variables in the console of VSCODE
`$ENV:LIBCLANG_PATH="H:\llvm\bin"` and `$ENV:FFMPEG_DIR="H:\ffmpeg-n4.3.1-26-gca55240b8c-win64-gpl-shared-4.3"` in the local vscode console. And this is the output I got by using `cargo build -vv --example dump-frames > output.txt`
[output.txt](https://github.com/zmwangx/rust-ffmpeg/files/5740854/output.txt)
and this is the console output I got that wasn't redirected to `output.txt`
[output_from_run.txt](https://github.com/zmwangx/rust-ffmpeg/files/5740856/output_from_run.txt)
I guess this means everything works! sorry for the trouble.
